parse.y

```diff
@@ -2613,6 +2614,64 @@ rb_str_to_parser_string(rb_parser_t *p, VALUE str)
     /* Type check */
     return rb_parser_encoding_string_new(p, RSTRING_PTR(str), RSTRING_LEN(str), rb_enc_get(str));
 }
+
+
+static void
+rb_parser_ary_extend(rb_parser_t *p, rb_parser_ary_t *ary, long len)
+{
+    long i;
+    if (ary->capa < len) {
+        ary->capa = len;
+        if (ary->data == NULL) {
+            ary->data = xcalloc(len, sizeof(void *));
+        } else {
+            ary->data = xrealloc(ary->data, sizeof(void *) * len);
+        }
+        for (i = ary->len; i < len; i++) {
+            ary->data[i] = Qnil;
+        }
+    }
+}
+
+static rb_parser_ary_t *
+rb_parser_ary_new_capa(rb_parser_t *p, long len)
+{
+    if (len < 0) {
+        rb_bug("negative array size (or size too big): %ld", len);
+    }
+    rb_parser_ary_t *ary = xcalloc(1, sizeof(rb_parser_ary_t));
+    long i;
+    ary->len = 0;
+    ary->capa = len;
+    if (0 < len) {
+        ary->data = xcalloc(len, sizeof(void *));
+    }
+    else {
+        ary->data = NULL;
+    }
+    for (i = 0; i < len; i++) {
+        ary->data[i] = Qnil;
+    }
+    return ary;
+}
+#define rb_parser_ary_new2 rb_parser_ary_new_capa
+
+inline static rb_parser_ary_t *
+rb_parser_ary_new(rb_parser_t *p)
+{
+    return rb_parser_ary_new_capa(p, 0);
+}
+
+static rb_parser_ary_t *
+rb_parser_ary_push(rb_parser_t *p, rb_parser_ary_t *ary, VALUE val)
+{
+    if (ary->len == ary->capa) {
+        rb_parser_ary_extend(p, ary, ary->len + 1);
```

Comment:
Expanding the array with `ary->len + 1` is no efficient especially for ast token array because it's expected to be expanded again and again. How about to double the capacity?

Reply:
Fixed: https://github.com/ruby/ruby/pull/10004/commits/d1fd5bae7fa1768ab5f179bd3c679996636c2b3e
BTW, what does the comment `// TODO` in `rb_ruby_parser_keep_tokens()` mean?

Reply:
Remove the line https://github.com/hasumikin/ruby/commit/64d18710129f1652a1ee5ea9d13bb168a4c5922d